src/Storages/StorageS3.cpp

```diff
@@ -356,6 +340,38 @@ BlockOutputStreamPtr StorageS3::write(const ASTPtr & /*query*/, const StorageMet
         max_single_part_upload_size);
 }
 
+void StorageS3::updateAuthSettings(const Context & context)
+{
+    auto settings = context.getStorageS3Settings().getSettings(uri.uri.toString());
+    if (client && (!access_key_id.empty() || settings == auth_settings))
```

Comment:
Can this place be a bottle neck? Comparing settings requires to compare several strings and we perform it in each `StorageS3::read` call.

Reply:
Decided to leave as is, because `StorageS3::read` called not often, once for query, I suppose